meta-balena-common/recipes-support/resin-init/resin-init-flasher/resin-init-flasher

```diff
@@ -133,26 +133,78 @@ if [ -z "$internal_dev" ]; then
     fail "Failed to find any block devices in $INTERNAL_DEVICE_KERNEL."
 fi
 inform "$internal_dev will be used for flashing."
-dd if="/opt/$BALENA_IMAGE" of="/dev/$internal_dev" bs=4M 2> /tmp/dd_progress_log & DD_PID=$!
 
-if ! kill -0 $DD_PID; then
-    fail "Failed to flash internal device $INTERNAL_DEVICE_KERNEL."
-fi
+if [ "$LUKS" = "1" ]; then
+    # TODO: Hardcoded password now, randomize with TPM2
+    PASSWORD="${LUKS_PASSWORD:-balenabeta}"
+    LOOP_DEVICE=$(losetup --find --show --partscan "/opt/$BALENA_IMAGE")
+
+    # Copy header and resin-boot partition
+    inform "Flashing boot partition"
+    dd if="/opt/$BALENA_IMAGE" of="/dev/$internal_dev" bs=4M count=11
+
+    # Repartition the new drive - this assumes GPT partition table
+    inform "Repartitioning $internal_dev for disk encryption"
+
+    echo -e "OK\nFix" | parted ---pretend-input-tty "/dev/$internal_dev" print
+
+    parted -s "/dev/$internal_dev" -- rm 5
+    parted -s "/dev/$internal_dev" -- rm 4
+    parted -s "/dev/$internal_dev" -- rm 3
+    parted -s "/dev/$internal_dev" -- rm 2
+    parted -s "/dev/$internal_dev" -- unit MiB mkpart resin-boot 44 86
+    parted -s "/dev/$internal_dev" -- unit MiB mkpart resin-rootA 86 1112
+    parted -s "/dev/$internal_dev" -- unit MiB mkpart resin-rootB 1112 2138
+    parted -s "/dev/$internal_dev" -- unit MiB mkpart resin-state 2138 2160
+    parted -s "/dev/$internal_dev" -- unit MiB mkpart resin-data 2160 2366
```

Comment:
Again, hardcoding partitioni sizes is not generic enough. The `image_types_balena` bbclass defines variables for the different partition sizes. At least the above should be based on those variables so that they can vary for different DTs.

Reply:
Changed